cfgrammar/src/lib/analysis.rs

```diff
@@ -0,0 +1,73 @@
+use crate::yacc::ast::{GrammarAST, Symbol};
+use crate::yacc::{YaccGrammarWarning, YaccGrammarWarningKind};
+use lazy_static::lazy_static;
+use std::ops::Deref;
+
+/// Performs an analysis on a given `Subject`, the given
+/// subject may be known to have pre-existing errors.
+pub trait Analysis<Subject> {
+    /// Perform an analysis on a given subject The mechanisms by which you retrieve the
+    /// results of an analysis are not specified by the trait, and particular to the types
+    /// that implement the trait.
+    fn analyze(&mut self, subject: &Subject);
+}
+
+pub struct YaccGrammarWarningAnalysis<SourceId>
+where
+    SourceId: PartialEq + ToOwned + ?Sized,
+{
+    src_id: SourceId::Owned,
+    warnings: Option<Box<[YaccGrammarWarning]>>,
+}
+
+impl<SourceId: PartialEq + ToOwned + ?Sized> YaccGrammarWarningAnalysis<SourceId> {
+    /// Create a new analysis given a SourceId such as the filename of a yacc file
+    /// For example, `YaccGrammarWarningAnalysis::new("foo.y")`
+    pub fn new(src_id: &SourceId) -> Self {
+        Self {
+            src_id: src_id.to_owned(),
+            warnings: None,
+        }
+    }
+    pub fn source_id(&self) -> &SourceId::Owned {
+        &self.src_id
+    }
+}
+
+impl<SourceId: PartialEq + ToOwned + ?Sized> Deref for YaccGrammarWarningAnalysis<SourceId> {
+    type Target = Box<[YaccGrammarWarning]>;
+
+    fn deref(&self) -> &Self::Target {
+        if let Some(warnings) = self.warnings.as_ref() {
+            warnings
+        } else {
+            lazy_static! {
+                static ref EMPTY_WARNINGS: Box<[YaccGrammarWarning]> = vec![].into_boxed_slice();
```

Comment:
This is... rather evil ;) I'd much rather have `warnings: Option<Vec<YaccGrammarWarning>>` and remove the `lazy_static`.

Reply:
> I'm not quite sure what the purpose of [1af9ba2](https://github.com/softdevteam/grmtools/commit/1af9ba262e001a6183f098102437aa6a56ef256a) -- it doesn't really seem related to this PR?

The reason it is here, is so that `build_grammar` can return a `YaccGrammarError` or a `Spanned` error.
We have Non-Spanned errors before that, (reading yacc files to build the AST), and non-spanned errors again after `build_grammar`  (`write_parser` here).

Reply:
Yeah, I couldn't get myself to expose the Option, because as long as an analysis is run it will be at least `Some(&[])`
So, this is just some weird case where the user allocated an analysis never ran it, and now wants to deref it.

I think the right thing here just not delay the allocation until `analyse` is called, but allocate `warnings` in `new` and drop the option entirely.

Reply:
should be fixed in [c2dd0fa](https://github.com/softdevteam/grmtools/pull/353/commits/c2dd0faef76b415da366c6a13a59fec9de2851b6)